### ISSUE TITLE
Secure Transport is deprecated.

### ISSUE THREAD

**Issue by kitanic** [sfackler/rust-native-tls#310]
native-tls on Apple platform uses Secure Transport via security-framework crate, however it is already [deprecated and unsupported](https://developer.apple.com/documentation/security/secure-transport).

Apple is asking for a move to the [Network](https://developer.apple.com/documentation/network) framework instead. I don't know if it is in this crate or in a sub-crate, implementation must be migrated.

**Comment by mmastrac** [comment]
Agreed. I've been seeing numerous crash and bug reports coming from Secure Transport -- it randomly throws permission errors for no discernable reason (possibly JAMF), and a user with nix installed is seeing Secure Transport in a crash report.

**Comment by sfackler** [comment]
How would that be implemented in a way that conforms to the native-tls APIs? In particular, the implementation needs to support a arbitrary stream under the TLS implementation.

**Comment by mmastrac** [comment]
@sfackler If possible, could we get native openssl support on MacOS instead?

**Comment by sfackler** [comment]
It's possible, but not really "native" then anymore.

**Comment by sfackler** [comment]
IIRC Apple does not want anyone linking to that and stopped shipping headers back when they switched from OpenSSL 0.9.8 to LibreSSL.

**Comment by mmastrac** [comment]
I tried to understand the state of LibreSSL support -- apparently Apple stopped shipping the OpenSSL headers _but_ continued to update and maintain the LibreSSL binary itself.

TBH it's not clear what the state of this library is. It's possible that it's stable, but it's also possible that it exists purely to satisfy older MacOS apps and scripts that expect the openssl command-line tool.

https://github.com/drduh/macOS-Security-and-Privacy-Guide/issues/356

OTOH, the Network.framework looks somewhat challenging to program against but not necessarily a bad fit.

**Comment by DemiMarie** [comment]
> How would that be implemented in a way that conforms to the native-tls APIs? In particular, the implementation needs to support a arbitrary stream under the TLS implementation.

Would it be possible to have a “with I/O” mode in which `Network.framework` handles not only TLS but everything below it?  This is the only solution for watchOS.

**Comment by DemiMarie** [comment]
@mmastrac @kitanic would either of you be willing to send patches?
